x-pack/libbeat/management/generate.go

```diff
@@ -0,0 +1,253 @@
+		}
+		return configList, nil
+	}
+
+	return r.transformFunc(cfg, agentInfo)
+}
+
+// ===========
+// Stream and Input processors
+// ===========
+
+// CreateInputsFromStreams breaks down the raw Expected config into an array of individual inputs/modules from the Streams values
+// that can later be formatted into the reloader's ConfigWithMetaData and sent to an indvidual beat/
+// This also performs the basic task of inserting module-level add_field processors into the inputs/modules.
+func CreateInputsFromStreams(raw *proto.UnitExpectedConfig, inputType string, agentInfo *client.AgentInfo) ([]map[string]interface{}, error) {
+	inputs := make([]map[string]interface{}, len(raw.Streams))
+
+	for iter := range raw.Streams {
+		streamSource := raw.Streams[iter].Source.AsMap()
+		streamSource = injectIndexStream(raw, inputType, iter, streamSource)
+		streamSource, err := injectStreamProcessors(raw, inputType, iter, streamSource)
+		if err != nil {
+			return nil, fmt.Errorf("Error injecting stream processors: %w", err)
+		}
+		streamSource, err = injectAgentInfoRule(streamSource, agentInfo)
+		if err != nil {
+			return nil, fmt.Errorf("Error injecting agent processors: %w", err)
+		}
+		inputs[iter] = streamSource
+	}
+
+	return inputs, nil
+}
+
+// CreateReloadConfigFromInputs turns a raw input/module list into the ConfigWithMeta type used by the reloader interface
+func CreateReloadConfigFromInputs(raw []map[string]interface{}) ([]*reload.ConfigWithMeta, error) {
+	// format for the reloadable list needed bythe cm.Reload() method
+	configList := make([]*reload.ConfigWithMeta, len(raw))
+
+	for iter := range raw {
+		uconfig, err := conf.NewConfigFrom(raw[iter])
+		if err != nil {
+			return nil, fmt.Errorf("error in conversion to conf.C: %w", err)
+		}
+		configList[iter] = &reload.ConfigWithMeta{Config: uconfig}
+	}
+	return configList, nil
+}
+
+// Emulates the InjectAgentInfoRule and InjectHeadersRule ast rules
+func injectAgentInfoRule(inputs map[string]interface{}, agentInfo *client.AgentInfo) (map[string]interface{}, error) {
+	// upstream API can sometimes return a nil agent info
+	if agentInfo == nil {
+		return inputs, nil
+	}
+	var processors = []interface{}{}
+
+	processors = append(processors, generateAddFieldsProcessor(
+		mapstr.M{"id": agentInfo.ID, "snapshot": agentInfo.Snapshot, "version": agentInfo.Version},
+		"elastic_agent"))
+	processors = append(processors, generateAddFieldsProcessor(
+		mapstr.M{"id": agentInfo.ID},
+		"agent"))
+
+	currentProcs, ok := inputs["processors"]
+	if !ok {
+		inputs["processors"] = processors
+	} else {
+		currentProcsList, ok := currentProcs.([]interface{})
+		if !ok {
+			return nil, fmt.Errorf("error creating list of existing processors, got: %#v", currentProcs)
+		}
+		inputs["processors"] = append(processors, currentProcsList...)
+
+	}
+
+	return inputs, nil
+}
+
+// injectIndexStream is an emulation of the InjectIndexProcessor AST code
```

Comment:
I haven't seen you explain what the AST is anywhere which will be hard for developers that know nothing about the V1 implementation to follow in the future.

Could you add a small README to the management package explaining what it does and what the history was?

Reply:
That's actually in my checklist! Just didn't want to have the PR wait while I worked through everything I wanted to write down.